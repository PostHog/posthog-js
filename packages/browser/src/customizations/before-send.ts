import { clampToRange } from '../utils/number-utils'
import { BeforeSendFn, CaptureResult, KnownEventName } from '../types'
import { includes } from '../utils/string-utils'
import { appendArray, sampleOnProperty, updateThreshold } from '../extensions/sampling'

/**
 * Provides an implementation of sampling that samples based on the distinct ID.
 * Using the provided percentage.
 * Can be used to create a beforeCapture fn for a PostHog instance.
 *
 * Setting 0.5 will cause roughly 50% of distinct ids to have events sent.
 * Not 50% of events for each distinct id.
 *
 * @param percent a number from 0 to 1, 1 means always send and, 0 means never send the event
 */
export function sampleByDistinctId(percent: number): BeforeSendFn {
    return (captureResult: CaptureResult | null): CaptureResult | null => {
        if (!captureResult) {
            return null
        }

        return sampleOnProperty(captureResult.properties.distinct_id, percent)
            ? {
                  ...captureResult,
                  properties: {
                      ...captureResult.properties,
                      $sample_type: ['sampleByDistinctId'],
                      $sample_threshold: percent,
                  },
              }
            : null
    }
}

/**
 * Provides an implementation of sampling that samples based on the session ID.
 * Using the provided percentage.
 * Can be used to create a beforeCapture fn for a PostHog instance.
 *
 * Setting 0.5 will cause roughly 50% of sessions to have events sent.
 * Not 50% of events for each session.
 *
 * @param percent a number from 0 to 1, 1 means always send and, 0 means never send the event
 */
export function sampleBySessionId(percent: number): BeforeSendFn {
    return (captureResult: CaptureResult | null): CaptureResult | null => {
        if (!captureResult) {
            return null
        }

        return sampleOnProperty(captureResult.properties.$session_id, percent)
            ? {
                  ...captureResult,
                  properties: {
                      ...captureResult.properties,
                      $sample_type: appendArray(captureResult.properties.$sample_type, 'sampleBySessionId'),
                      $sample_threshold: updateThreshold(captureResult.properties.$sample_threshold, percent),
                  },
              }
            : null
    }
}

/**
 * Provides an implementation of sampling that samples based on the event name.
 * Using the provided percentage.
 * Can be used to create a beforeCapture fn for a PostHog instance.
 *
 * @param eventNames an array of event names to sample, sampling is applied across events not per event name
 * @param percent a number from 0 to 1, 1 means always send, 0 means never send the event
 */
export function sampleByEvent(eventNames: KnownEventName[], percent: number): BeforeSendFn {
    return (captureResult: CaptureResult | null): CaptureResult | null => {
        if (!captureResult) {
            return null
        }

        if (!includes(eventNames, captureResult.event)) {
            return captureResult
        }

        const number = Math.random()
        return number * 100 < clampToRange(percent * 100, 0, 100)
            ? {
                  ...captureResult,
                  properties: {
                      ...captureResult.properties,
                      $sample_type: appendArray(captureResult.properties?.$sample_type, 'sampleByEvent'),
                      $sample_threshold: updateThreshold(captureResult.properties?.$sample_threshold, percent),
                      $sampled_events: appendArray(captureResult.properties?.$sampled_events, eventNames),
                  },
              }
            : null
    }
}

export const printAndDropEverything: BeforeSendFn = (result) => {
    // eslint-disable-next-line no-console
    console.log('Would have sent event:', result)
    return null
}
