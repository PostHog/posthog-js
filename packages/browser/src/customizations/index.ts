export * from './setAllPersonProfilePropertiesAsPersonPropertiesForFlags'
export * from './before-send'
export {
    posthogReduxLogger,
    posthogKeaLogger,
    sessionRecordingLoggerForPostHogInstance,
    browserConsoleLogger,
} from './posthogReduxLogger'
