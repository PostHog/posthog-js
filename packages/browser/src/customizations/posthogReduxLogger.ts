import { BucketedRateLimiter, isNullish, isObject, isUndefined, Logger } from '@posthog/core'
import { createLogger } from '../utils/logger'
import type { PostHog } from '../posthog-core'

// types copied from redux toolkit so we can avoid taking a dependency in the library and confusing people using the SDK

/**
 * An *unknown* action.
 * This is the most minimal possible shape for an action.
 * Allows for type-safe usage of actions without dependencies.
 */
export interface UnknownAction {
    type: string
    [extraProps: string]: unknown
}

/**
 * A *dispatching function* (or simply *dispatch function*) is a function that
 * accepts an action or an async action; it then may or may not dispatch one
 * or more actions to the store.
 */
export interface Dispatch<A extends UnknownAction = UnknownAction> {
    <T extends A>(action: T): T
}

/**
 * A middleware is a higher-order function that composes a dispatch function
 * to return a new dispatch function. It often turns async actions into
 * actions.
 *
 * This matches Redux Toolkit's Middleware interface for compatibility.
 */
export interface ReduxMiddleware<
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-object-type
    _DispatchExt = {},
    S = any,
    D extends Dispatch = Dispatch,
> {
    (api: MiddlewareAPI<D, S>): (next: (action: unknown) => unknown) => (action: unknown) => unknown
}

/**
 * A middleware API is an object containing the store's dispatch function and getState function.
 * A middleware is given the middleware API as its first parameter.
 */
export interface MiddlewareAPI<D extends Dispatch = Dispatch, S = any> {
    dispatch: D
    getState(): S
}

// end of copied types

export interface StateEvent {
    type: string
    payload?: any
    timestamp: number
    executionTimeMs?: number
    prevState: any
    nextState: any
}

export type PostHogStateLogger = (title: string, stateEvent: StateEvent) => void

export interface PostHogStateLoggerConfig<S = any> {
    maskAction?: (action: UnknownAction) => UnknownAction | null
    maskState?: (state: S, action?: UnknownAction) => S
    titleFunction?: (stateEvent: StateEvent) => string
    logger?: PostHogStateLogger
    diffState?: boolean
    /**
     * actions logging is token bucket rate limited to avoid flooding
     * this controls the rate limiter's refill rate, see BucketedRateLimiter docs for details
     * normally this is only changed with posthog support assistance
     */
    rateLimiterRefillRate?: number
    /**
     * actions logging is token bucket rate limited to avoid flooding
     * this controls the rate limiter's bucket size, see BucketedRateLimiter docs for details
     * normally this is only changed with posthog support assistance
     */
    rateLimiterBucketSize?: number
    /**
     * Controls how many levels deep the state diffing goes when `diffState` is true
     * Defaults to 10
     * Increase this if you have nested state changes that are not being captured
     * Decrease this if you are seeing too much state in the diffs and want to reduce noise
     * Note that increasing this will increase the CPU cost of diffing, so use with caution
     * and only increase if necessary
     * Normally this is only changed with posthog support assistance
     */
    __stateComparisonDepth?: number
}

/**
 * Default title function for Redux events
 */
function defaultTitleFunction(stateEvent: StateEvent): string {
    const { type, executionTimeMs } = stateEvent
    const timeText = isNullish(executionTimeMs) ? '' : ` (${executionTimeMs.toFixed(2)}ms)`
    return `${type}${timeText}`
}

// we need a posthog logger for the rate limiter
const phConsoleLogger: Logger = createLogger('[PostHog Action RateLimiting]')

export function browserConsoleLogger(title: string, stateEvent: StateEvent): void {
    // but the posthog logger swallows messages unless debug is on
    // so we don't want to use it in this default logger
    // eslint-disable-next-line no-console
    console.log(title, stateEvent)
}

/**
 * Logger that sends state events to PostHog session recordings
 * Requires that the loaded posthog instance is provided
 * And returns the function to use as the logger
 *
 * e.g. const config = { logger: sessionRecordingLoggerForPostHogInstance(posthog) }
 */
export const sessionRecordingLoggerForPostHogInstance: (posthog: PostHog) => PostHogStateLogger =
    (postHogInstance: PostHog) =>
    (title: string, stateEvent: StateEvent): void => {
        postHogInstance?.sessionRecording?.tryAddCustomEvent('app-state', { title, stateEvent })
    }

/**
 * Get only the changed keys from two states
 * NB exported for testing purposes only, not part of the public API and may change without warning
 *
 * Returns { prevState: changedKeysOnly, nextState: changedKeysOnly }
 */
export function getChangedStateKeys<S>(
    prevState: S,
    nextState: S,
    maxDepth: number = 3
): { prevState?: Partial<S>; nextState?: Partial<S> } {
    // Fast bailouts
    if (prevState === nextState) return { prevState: {}, nextState: {} }
    if (!prevState || !nextState) return { prevState: prevState as Partial<S>, nextState: nextState as Partial<S> }
    if (typeof prevState !== 'object' || typeof nextState !== 'object') return {}
    if (maxDepth <= 0) return { prevState: prevState as Partial<S>, nextState: nextState as Partial<S> }

    const prev: any = {}
    const next: any = {}
    let hasChanges = false

    // Single pass: check all keys from both objects
    const allKeys = new Set([...Object.keys(prevState), ...Object.keys(nextState)])

    for (const key of allKeys) {
        const prevVal = (prevState as any)[key]
        const nextVal = (nextState as any)[key]

        if (prevVal === nextVal) continue // No change

        hasChanges = true

        // Handle missing keys
        if (isUndefined(prevVal)) {
            next[key] = nextVal
            continue
        }
        if (isUndefined(nextVal)) {
            prev[key] = prevVal
            continue
        }

        // For objects, recurse once more, otherwise treat as primitive
        if (isObject(prevVal) && isObject(nextVal)) {
            const nested = getChangedStateKeys(prevVal, nextVal, maxDepth - 1)

            // Only include if there are actual changes
            if (nested.prevState && Object.keys(nested.prevState).length > 0) prev[key] = nested.prevState
            if (nested.nextState && Object.keys(nested.nextState).length > 0) next[key] = nested.nextState
        } else {
            // Primitive or array - include both values
            prev[key] = prevVal
            next[key] = nextVal
        }
    }

    return hasChanges ? { prevState: prev, nextState: next } : { prevState: {}, nextState: {} }
}

// Debounced logger for rate limit messages
const createDebouncedActionRateLimitedLogger = () => {
    let timeout: ReturnType<typeof setTimeout> | null = null
    let ignoredCount = 0
    let lastActionType: string | null = null

    return {
        info: (actionType: string) => {
            if (lastActionType !== actionType) {
                // Reset counter when action type changes
                ignoredCount = 0
                lastActionType = actionType
            }

            ignoredCount++

            if (timeout) {
                clearTimeout(timeout)
            }

            timeout = setTimeout(() => {
                const count = ignoredCount

                if (count === 1) {
                    phConsoleLogger.info(`action "${actionType}" has been rate limited`)
                } else {
                    phConsoleLogger.info(`action "${actionType}" has been rate limited (${count} times)`)
                }

                ignoredCount = 0
                timeout = null
            }, 1000)
        },
    }
}

const debouncedActionRateLimitedLogger = createDebouncedActionRateLimitedLogger()

/**
 * Creates a Kea plugin that logs actions and state changes to a provided logger
 * This can be used as a plugin in any Kea setup to capture state changes
 */
export function posthogKeaLogger<S = any>(config: PostHogStateLoggerConfig<S> = {}) {
    const middleware = posthogReduxLogger(config)

    return {
        name: 'posthog-kea-logger',
        events: {
            beforeReduxStore(options: any) {
                options.middleware.push(middleware)
            },
        },
    }
}

/**
 * Creates a Redux middleware that logs actions and state changes to a provided logger
 * This can be used as middleware in any Redux store to capture state changes
 *
 * The logging uses token-bucket rate limiting to avoid flooding the logging with many changes
 * by default logging rate limiting captures ten action instances before rate limiting by action type
 * refills at a rate of one token / 1-second period
 * e.g. will capture 1 rate limited action every 1 second until the burst ends
 */
export function posthogReduxLogger<S = any>(
    config: PostHogStateLoggerConfig<S> = {}
    // the empty object is the recommended typing from redux docs
    //eslint-disable-next-line @typescript-eslint/no-empty-object-type
): ReduxMiddleware<{}, S> {
    const {
        maskAction,
        maskState,
        titleFunction = defaultTitleFunction,
        logger = browserConsoleLogger,
        diffState = true,
        rateLimiterRefillRate = 1,
        rateLimiterBucketSize = 10,
        __stateComparisonDepth,
    } = config

    const rateLimiter: BucketedRateLimiter<string> = new BucketedRateLimiter({
        refillRate: rateLimiterRefillRate,
        bucketSize: rateLimiterBucketSize,
        refillInterval: 1000, // one second in milliseconds,
        _logger: phConsoleLogger,
    })

    return (store: MiddlewareAPI<Dispatch, S>) =>
        (next: (action: unknown) => unknown) =>
        (action: unknown): unknown => {
            const typedAction = action as UnknownAction
            // Get the state before the action
            const prevState = store.getState()

            // Track execution time
            // eslint-disable-next-line compat/compat
            const startTime = performance.now()

            const result = next(typedAction)

            // eslint-disable-next-line compat/compat
            const endTime = performance.now()
            const executionTimeMs = endTime - startTime

            // Get the state after the action
            const nextState = store.getState()

            const maskedAction = maskAction ? maskAction(typedAction) : typedAction

            if (!maskedAction) {
                return result
            }

            const isRateLimited = rateLimiter.consumeRateLimit(typedAction.type)

            if (isRateLimited) {
                debouncedActionRateLimitedLogger.info(typedAction.type)
            } else {
                // Apply masking to states
                try {
                    const maskedPrevState = maskState ? maskState(prevState, maskedAction) : prevState
                    const maskedNextState = maskState ? maskState(nextState, maskedAction) : nextState

                    type FilteredState = { 'invalid state'?: string } | (S & Record<string, any>)
                    let filteredPrevState: FilteredState
                    let filteredNextState: FilteredState
                    if (diffState) {
                        const { prevState: diffedPrevState, nextState: diffedNextState } = getChangedStateKeys(
                            maskedPrevState,
                            maskedNextState,
                            __stateComparisonDepth ?? 10
                        )
                        const invalidPayloadForDiffing: FilteredState = { 'invalid state': 'no changes after diffing' }
                        filteredPrevState = (diffedPrevState as FilteredState) ?? invalidPayloadForDiffing
                        filteredNextState = (diffedNextState as FilteredState) ?? invalidPayloadForDiffing
                    } else {
                        const invalidPayloadForLogging = { 'invalid state': 'logger only supports object payloads' }
                        filteredPrevState = isObject(maskedPrevState) ? maskedPrevState : invalidPayloadForLogging
                        filteredNextState = isObject(maskedNextState) ? maskedNextState : invalidPayloadForLogging
                    }

                    const { type, ...actionData } = maskedAction

                    const reduxEvent: StateEvent = {
                        type,
                        payload: actionData,
                        timestamp: Date.now(),
                        executionTimeMs,
                        prevState: filteredPrevState,
                        nextState: filteredNextState,
                    }

                    const title = titleFunction(reduxEvent)
                    logger(title, reduxEvent)
                } catch (e: any) {
                    // logging should never throw errors and break someone's app
                    phConsoleLogger.error('Error logging state:', e)
                }
            }

            return result
        }
}
