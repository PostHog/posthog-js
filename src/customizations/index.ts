export * from './setAllPersonProfilePropertiesAsPersonPropertiesForFlags'
export * from './before-send'
