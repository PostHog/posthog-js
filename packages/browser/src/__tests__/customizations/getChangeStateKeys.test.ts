import { getChangedStateKeys } from '../../customizations/posthogReduxLogger'

// Helper function to create complex nested state objects
function createComplexState(depth: number, breadth: number, includeArrays = false): Record<string, any> {
    const state: Record<string, any> = {}

    for (let i = 0; i < breadth; i++) {
        const key = `key_${i}`

        if (depth > 0) {
            // Create nested objects
            state[key] = createComplexState(depth - 1, breadth, includeArrays)
        } else {
            // Create leaf values
            if (includeArrays && i % 3 === 0) {
                state[key] = new Array(10).fill(0).map((_, idx) => ({ id: idx, value: `item_${idx}` }))
            } else if (i % 2 === 0) {
                state[key] = `value_${i}_${Math.random().toString(36).substr(2, 9)}`
            } else {
                state[key] = Math.floor(Math.random() * 1000)
            }
        }
    }

    return state
}

// Helper function to modify state in a realistic way (simulating UI drag operations)
function modifyStateForDrag(state: Record<string, any>, modifications: number = 5): Record<string, any> {
    const newState = JSON.parse(JSON.stringify(state))

    // Simulate drag-related state changes
    for (let i = 0; i < modifications; i++) {
        const key = `drag_operation_${i}`
        newState[key] = {
            isDragging: true,
            dragStartTime: Date.now(),
            dragPosition: { x: Math.random() * 1000, y: Math.random() * 1000 },
            dragElement: `element_${i}`,
            dragData: new Array(20).fill(0).map((_, idx) => ({
                id: idx,
                position: { x: idx * 10, y: idx * 10 },
                metadata: { type: 'draggable', visible: true },
            })),
        }
    }

    // Modify some existing deep properties
    if (newState.key_0 && typeof newState.key_0 === 'object') {
        Object.keys(newState.key_0).forEach((key, idx) => {
            if (idx < modifications) {
                if (typeof newState.key_0[key] === 'object') {
                    newState.key_0[key] = { ...newState.key_0[key], modified: true, timestamp: Date.now() }
                }
            }
        })
    }

    return newState
}

describe('getChangeStateKeys', () => {
    describe('performance tests', () => {
        test('should handle realistic UI state efficiently', () => {
            // Fixed test data representing realistic UI state
            const prevState = {
                ui: {
                    modal: { isOpen: false, content: null },
                    sidebar: { collapsed: false, activeTab: 'files' },
                    editor: {
                        content: 'hello world',
                        cursor: { line: 1, column: 5 },
                        selection: null,
                    },
                },
                data: {
                    files: [
                        { id: 1, name: 'file1.js', modified: false },
                        { id: 2, name: 'file2.js', modified: false },
                    ],
                    user: { id: 123, name: 'John', preferences: { theme: 'dark' } },
                },
            }

            const nextState = {
                ...prevState,
                ui: {
                    ...prevState.ui,
                    modal: { isOpen: true, content: 'Save changes?' },
                    editor: {
                        ...prevState.ui.editor,
                        cursor: { line: 1, column: 11 },
                        selection: { start: 6, end: 11 },
                    },
                },
            }

            const startTime = performance.now()
            const result = getChangedStateKeys(prevState, nextState)
            const endTime = performance.now()

            const executionTime = endTime - startTime

            // Performance assertion - should complete within 5ms for realistic UI state
            expect(executionTime).toBeLessThan(5)

            // Correctness assertions - should detect UI changes
            expect(result.nextState).toHaveProperty('ui')
            expect(result.nextState.ui).toHaveProperty('modal')
            expect(result.nextState.ui).toHaveProperty('editor')

            console.log(`Realistic UI state diff took: ${executionTime.toFixed(2)}ms`)
        })

        test('should handle medium complexity state objects', () => {
            const prevState = createComplexState(3, 8, true)
            const nextState = modifyStateForDrag(prevState, 5)

            const startTime = performance.now()
            const result = getChangedStateKeys(prevState, nextState)
            const endTime = performance.now()

            const executionTime = endTime - startTime

            expect(executionTime).toBeLessThan(30)

            // Correctness assertions
            expect(result.nextState).toBeDefined()
            expect(Object.keys(result.nextState || {})).toEqual(
                expect.arrayContaining(['drag_operation_0', 'drag_operation_1', 'drag_operation_2'])
            )

            console.log(`Medium state diff took: ${executionTime.toFixed(2)}ms`)
        })

        test('should handle large complex state objects (stress test)', () => {
            const prevState = createComplexState(4, 10, true)
            const nextState = modifyStateForDrag(prevState, 8)

            const startTime = performance.now()
            const result = getChangedStateKeys(prevState, nextState)
            const endTime = performance.now()

            const executionTime = endTime - startTime

            // Performance assertion - should complete quickly even for large objects
            // This is the threshold where UI lag would become noticeable during drag operations
            expect(executionTime).toBeLessThan(30)

            // Correctness assertions
            expect(result.nextState).toBeDefined()
            expect(Object.keys(result.nextState || {}).length).toBeGreaterThan(5)

            console.log(`Large state diff took: ${executionTime.toFixed(2)}ms`)
            console.log(`State size: ~${JSON.stringify(prevState).length} characters`)
        })

        test('should handle complex state changes efficiently', () => {
            // Fixed complex state - easier to understand what we're testing
            const prevState = {
                app: {
                    user: { id: 1, name: 'John', settings: { theme: 'dark', lang: 'en' } },
                    data: {
                        items: [
                            { id: 1, text: 'Item 1' },
                            { id: 2, text: 'Item 2' },
                        ],
                        filters: { active: true, category: 'all' },
                    },
                },
                ui: {
                    modal: { open: false, type: null },
                    notifications: { count: 0, items: [] },
                },
            }

            const nextState = {
                ...prevState,
                app: {
                    ...prevState.app,
                    user: { ...prevState.app.user, name: 'Jane' },
                    data: {
                        ...prevState.app.data,
                        items: [...prevState.app.data.items, { id: 3, text: 'Item 3' }],
                        filters: { ...prevState.app.data.filters, category: 'active' },
                    },
                },
                ui: {
                    ...prevState.ui,
                    modal: { open: true, type: 'confirm' },
                },
            }

            // Test performance with multiple runs
            const iterations = 10
            const times: number[] = []

            for (let i = 0; i < iterations; i++) {
                const startTime = performance.now()
                const result = getChangedStateKeys(prevState, nextState)
                const endTime = performance.now()
                times.push(endTime - startTime)

                // Verify correctness on first run
                if (i === 0) {
                    expect(result.nextState).toHaveProperty('app')
                    expect(result.nextState).toHaveProperty('ui')
                    expect(result.nextState.app).toHaveProperty('user')
                    expect(result.nextState.app).toHaveProperty('data')
                }
            }

            const avgTime = times.reduce((a, b) => a + b) / times.length
            const maxTime = Math.max(...times)

            // Should be fast for realistic complex state
            expect(avgTime).toBeLessThan(10)
            expect(maxTime).toBeLessThan(20)

            console.log(`Complex state diff - Avg: ${avgTime.toFixed(2)}ms, Max: ${maxTime.toFixed(2)}ms`)
        })

        test('should handle identical states efficiently', () => {
            const state = createComplexState(4, 8, true)

            const startTime = performance.now()
            const result = getChangedStateKeys(state, state)
            const endTime = performance.now()

            const executionTime = endTime - startTime

            // Should be very fast for identical states
            expect(executionTime).toBeLessThan(1)
            expect(result.prevState).toEqual({})
            expect(result.nextState).toEqual({})

            console.log(`Identical state diff took: ${executionTime.toFixed(2)}ms`)
        })
    })

    describe('correctness tests', () => {
        test('should return empty objects for identical states', () => {
            const state = { a: 1, b: { c: 2 } }
            const result = getChangedStateKeys(state, state)

            expect(result.prevState).toEqual({})
            expect(result.nextState).toEqual({})
        })

        test('should handle null and undefined values', () => {
            const prevState = { a: null, b: undefined, c: 'value' }
            const nextState = { a: 'changed', b: 'defined', c: null }

            const result = getChangedStateKeys(prevState, nextState)

            expect(result.prevState).toEqual({ a: null, b: undefined, c: 'value' })
            expect(result.nextState).toEqual({ a: 'changed', b: 'defined', c: null })
        })

        test('should correctly identify added keys', () => {
            const prevState = { a: 1, b: 2 }
            const nextState = { a: 1, b: 2, c: 3 }

            const result = getChangedStateKeys(prevState, nextState)

            expect(result.nextState).toEqual({ c: 3 })
            expect(result.prevState).toEqual({})
        })

        test('should correctly identify removed keys', () => {
            const prevState = { a: 1, b: 2, c: 3 }
            const nextState = { a: 1, b: 2 }

            const result = getChangedStateKeys(prevState, nextState)

            expect(result.prevState).toEqual({ c: 3 })
            expect(result.nextState).toEqual({})
        })

        test('should correctly identify changed keys', () => {
            const prevState = { a: 1, b: 2, c: 3 }
            const nextState = { a: 1, b: 5, c: 3 }

            const result = getChangedStateKeys(prevState, nextState)

            expect(result.prevState).toEqual({ b: 2 })
            expect(result.nextState).toEqual({ b: 5 })
        })

        test('should handle arrays as primitive values', () => {
            const prevState = {
                list: [1, 2, 3],
                obj: { a: 1 },
            }
            const nextState = {
                list: [1, 2, 3, 4],
                obj: { a: 1 },
            }

            const result = getChangedStateKeys(prevState, nextState)

            expect(result.prevState).toEqual({ list: [1, 2, 3] })
            expect(result.nextState).toEqual({ list: [1, 2, 3, 4] })
        })

        test('should handle nested object changes', () => {
            const prevState = {
                user: { name: 'John', age: 30 },
                settings: { theme: 'dark', notifications: { email: true } },
            }
            const nextState = {
                user: { name: 'John', age: 31 },
                settings: { theme: 'light', notifications: { email: true } },
            }

            const result = getChangedStateKeys(prevState, nextState)

            expect(result.prevState).toEqual({
                user: { age: 30 },
                settings: { theme: 'dark' },
            })
            expect(result.nextState).toEqual({
                user: { age: 31 },
                settings: { theme: 'light' },
            })
        })

        test('should handle deeply nested changes within depth limit', () => {
            const prevState = {
                level1: {
                    level2: {
                        level3: {
                            value: 'old',
                        },
                    },
                },
            }
            const nextState = {
                level1: {
                    level2: {
                        level3: {
                            value: 'new',
                        },
                    },
                },
            }

            const result = getChangedStateKeys(prevState, nextState)

            expect(result.prevState).toEqual({
                level1: {
                    level2: {
                        level3: { value: 'old' },
                    },
                },
            })
            expect(result.nextState).toEqual({
                level1: {
                    level2: {
                        level3: { value: 'new' },
                    },
                },
            })
        })

        test('should respect depth limit', () => {
            const prevState = {
                l1: { l2: { l3: { l4: { l5: 'old' } } } },
            }
            const nextState = {
                l1: { l2: { l3: { l4: { l5: 'new' } } } },
            }

            // With default maxDepth=3, should stop at level 3
            const result = getChangedStateKeys(prevState, nextState, 2)

            expect(result.prevState).toEqual({
                l1: { l2: { l3: { l4: { l5: 'old' } } } },
            })
            expect(result.nextState).toEqual({
                l1: { l2: { l3: { l4: { l5: 'new' } } } },
            })
        })

        test('should handle mixed data types', () => {
            const prevState = {
                string: 'hello',
                number: 42,
                boolean: true,
                array: [1, 2, 3],
                object: { nested: 'value' },
                nullValue: null,
                undefinedValue: undefined,
            }
            const nextState = {
                string: 'world',
                number: 42,
                boolean: false,
                array: [4, 5, 6],
                object: { nested: 'changed' },
                nullValue: 'not null',
                undefinedValue: 'defined',
            }

            const result = getChangedStateKeys(prevState, nextState)

            expect(result.prevState).toEqual({
                string: 'hello',
                boolean: true,
                array: [1, 2, 3],
                object: { nested: 'value' },
                nullValue: null,
                undefinedValue: undefined,
            })
            expect(result.nextState).toEqual({
                string: 'world',
                boolean: false,
                array: [4, 5, 6],
                object: { nested: 'changed' },
                nullValue: 'not null',
                undefinedValue: 'defined',
            })
        })

        test('should handle edge case - empty objects', () => {
            const prevState = {}
            const nextState = { a: 1 }

            const result = getChangedStateKeys(prevState, nextState)

            expect(result.prevState).toEqual({})
            expect(result.nextState).toEqual({ a: 1 })
        })

        test('should handle edge case - non-object inputs', () => {
            const result1 = getChangedStateKeys('string', { a: 1 })
            const result2 = getChangedStateKeys({ a: 1 }, null)
            const result3 = getChangedStateKeys(42, true)

            expect(result1).toEqual({})
            expect(result2).toEqual({ prevState: { a: 1 }, nextState: null })
            expect(result3).toEqual({})
        })

        test('should handle Redux-style state changes', () => {
            // Simulate a typical Redux state change
            const prevState = {
                todos: [
                    { id: 1, text: 'Learn React', completed: false },
                    { id: 2, text: 'Learn Redux', completed: true },
                ],
                visibilityFilter: 'SHOW_ALL',
                user: {
                    id: 123,
                    name: 'John',
                    profile: { avatar: 'old.jpg', theme: 'dark' },
                },
            }

            const nextState = {
                todos: [
                    { id: 1, text: 'Learn React', completed: true }, // completed changed
                    { id: 2, text: 'Learn Redux', completed: true },
                    { id: 3, text: 'Learn Testing', completed: false }, // new todo added
                ],
                visibilityFilter: 'SHOW_COMPLETED', // filter changed
                user: {
                    id: 123,
                    name: 'John',
                    profile: { avatar: 'new.jpg', theme: 'dark' }, // avatar changed
                },
            }

            const result = getChangedStateKeys(prevState, nextState)

            // Should detect array changes as primitive (entire array replaced)
            expect(result.prevState).toHaveProperty('todos')
            expect(result.nextState).toHaveProperty('todos')
            expect(result.prevState).toHaveProperty('visibilityFilter', 'SHOW_ALL')
            expect(result.nextState).toHaveProperty('visibilityFilter', 'SHOW_COMPLETED')
            expect(result.prevState.user).toHaveProperty('profile')
            expect(result.nextState.user).toHaveProperty('profile')
            expect(result.nextState.user.profile).toHaveProperty('avatar', 'new.jpg')
        })
    })
})
