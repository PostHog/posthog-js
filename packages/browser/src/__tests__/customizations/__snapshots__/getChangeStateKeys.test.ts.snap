// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getChangeStateKeys correctness tests should handle Redux-style state changes 1`] = `
Object {
  "nextState": Object {
    "todos": Array [
      Object {
        "completed": true,
        "id": 1,
        "text": "Learn React",
      },
      Object {
        "completed": true,
        "id": 2,
        "text": "Learn Redux",
      },
      Object {
        "completed": false,
        "id": 3,
        "text": "Learn Testing",
      },
    ],
    "user": Object {
      "profile": Object {
        "avatar": "new.jpg",
      },
    },
    "visibilityFilter": "SHOW_COMPLETED",
  },
  "prevState": Object {
    "todos": Array [
      Object {
        "completed": false,
        "id": 1,
        "text": "Learn React",
      },
      Object {
        "completed": true,
        "id": 2,
        "text": "Learn Redux",
      },
    ],
    "user": Object {
      "profile": Object {
        "avatar": "old.jpg",
      },
    },
    "visibilityFilter": "SHOW_ALL",
  },
}
`;

exports[`getChangeStateKeys correctness tests should handle added keys 1`] = `
Object {
  "nextState": Object {
    "c": 3,
  },
  "prevState": Object {},
}
`;

exports[`getChangeStateKeys correctness tests should handle arrays as primitive values 1`] = `
Object {
  "nextState": Object {
    "list": Array [
      1,
      2,
      3,
      4,
    ],
  },
  "prevState": Object {
    "list": Array [
      1,
      2,
      3,
    ],
  },
}
`;

exports[`getChangeStateKeys correctness tests should handle changed keys 1`] = `
Object {
  "nextState": Object {
    "b": 5,
  },
  "prevState": Object {
    "b": 2,
  },
}
`;

exports[`getChangeStateKeys correctness tests should handle deeply nested changes within depth limit 1`] = `
Object {
  "nextState": Object {
    "level1": Object {
      "level2": Object {
        "level3": Object {
          "value": "new",
        },
      },
    },
  },
  "prevState": Object {
    "level1": Object {
      "level2": Object {
        "level3": Object {
          "value": "old",
        },
      },
    },
  },
}
`;

exports[`getChangeStateKeys correctness tests should handle depth limit respected 1`] = `
Object {
  "nextState": Object {
    "l1": Object {
      "l2": Object {
        "l3": Object {
          "l4": Object {
            "l5": "new",
          },
        },
      },
    },
  },
  "prevState": Object {
    "l1": Object {
      "l2": Object {
        "l3": Object {
          "l4": Object {
            "l5": "old",
          },
        },
      },
    },
  },
}
`;

exports[`getChangeStateKeys correctness tests should handle empty objects 1`] = `
Object {
  "nextState": Object {
    "a": 1,
  },
  "prevState": Object {},
}
`;

exports[`getChangeStateKeys correctness tests should handle identical states 1`] = `
Object {
  "nextState": Object {},
  "prevState": Object {},
}
`;

exports[`getChangeStateKeys correctness tests should handle mixed data types 1`] = `
Object {
  "nextState": Object {
    "array": Array [
      4,
      5,
      6,
    ],
    "boolean": false,
    "nullValue": "not null",
    "object": Object {
      "nested": "changed",
    },
    "string": "world",
    "undefinedValue": "defined",
  },
  "prevState": Object {
    "array": Array [
      1,
      2,
      3,
    ],
    "boolean": true,
    "nullValue": null,
    "object": Object {
      "nested": "value",
    },
    "string": "hello",
  },
}
`;

exports[`getChangeStateKeys correctness tests should handle nested object changes 1`] = `
Object {
  "nextState": Object {
    "settings": Object {
      "theme": "light",
    },
    "user": Object {
      "age": 31,
    },
  },
  "prevState": Object {
    "settings": Object {
      "theme": "dark",
    },
    "user": Object {
      "age": 30,
    },
  },
}
`;

exports[`getChangeStateKeys correctness tests should handle null and undefined values 1`] = `
Object {
  "nextState": Object {
    "a": "changed",
    "b": "defined",
    "c": null,
  },
  "prevState": Object {
    "a": null,
    "c": "value",
  },
}
`;

exports[`getChangeStateKeys correctness tests should handle number vs boolean (non-object input) 1`] = `Object {}`;

exports[`getChangeStateKeys correctness tests should handle object vs null (non-object input) 1`] = `
Object {
  "nextState": null,
  "prevState": Object {
    "a": 1,
  },
}
`;

exports[`getChangeStateKeys correctness tests should handle removed keys 1`] = `
Object {
  "nextState": Object {},
  "prevState": Object {
    "c": 3,
  },
}
`;

exports[`getChangeStateKeys correctness tests should handle string vs object (non-object input) 1`] = `Object {}`;
